### DISCUSSION TITLE
`lime rebuild tools` fails with Haxe nightly

### DISCUSSION THREAD

**Discussion by tobil4sk** [openfl/lime#1645] [General]
Using the develop branch:

```
$ haxelib run lime rebuild tools
/usr/local/share/haxe/std/haxe/display/Protocol.hx:32: characters 2-105 : Inline variable initialization must be a constant value
/usr/local/share/haxe/std/haxe/display/Display.hx:39: characters 2-113 : Inline variable initialization must be a constant value
/usr/local/share/haxe/std/haxe/display/Display.hx:44: characters 2-158 : Inline variable initialization must be a constant value
/usr/local/share/haxe/std/haxe/display/Display.hx:49: characters 2-125 : Inline variable initialization must be a constant value
/usr/local/share/haxe/std/haxe/display/Display.hx:54: characters 2-119 : Inline variable initialization must be a constant value
/usr/local/share/haxe/std/haxe/display/Display.hx:59: characters 2-127 : Inline variable initialization must be a constant value
/usr/local/share/haxe/std/haxe/display/Display.hx:64: characters 2-131 : Inline variable initialization must be a constant value
/usr/local/share/haxe/std/haxe/display/Display.hx:69: characters 2-96 : Inline variable initialization must be a constant value
/usr/local/share/haxe/std/haxe/display/Display.hx:74: characters 2-116 : Inline variable initialization must be a constant value
/usr/local/share/haxe/std/haxe/display/Display.hx:79: characters 2-125 : Inline variable initialization must be a constant value
/usr/local/share/haxe/std/haxe/display/Protocol.hx:32: characters 2-105 : Inline variable initialization must be a constant value
/usr/local/share/haxe/std/haxe/display/Display.hx:39: characters 2-113 : Inline variable initialization must be a constant value
/usr/local/share/haxe/std/haxe/display/Display.hx:44: characters 2-158 : Inline variable initialization must be a constant value
/usr/local/share/haxe/std/haxe/display/Display.hx:49: characters 2-125 : Inline variable initialization must be a constant value
/usr/local/share/haxe/std/haxe/display/Display.hx:54: characters 2-119 : Inline variable initialization must be a constant value
/usr/local/share/haxe/std/haxe/display/Display.hx:59: characters 2-127 : Inline variable initialization must be a constant value
/usr/local/share/haxe/std/haxe/display/Display.hx:64: characters 2-131 : Inline variable initialization must be a constant value
/usr/local/share/haxe/std/haxe/display/Display.hx:69: characters 2-96 : Inline variable initialization must be a constant value
/usr/local/share/haxe/std/haxe/display/Display.hx:74: characters 2-116 : Inline variable initialization must be a constant value
/usr/local/share/haxe/std/haxe/display/Display.hx:79: characters 2-125 : Inline variable initialization must be a constant value
../src/lime/graphics/Image.hx:1053: characters 17-41 : Unknown<0> cannot be constructed
../src/lime/ui/Window.hx:56: characters 41-64 : Unknown<0> cannot be constructed
```

Runs with no issues on 4.2.5. Could be related to #1373?

**Comment by player-03** [comment]
Just making sure: is /usr/local/share/haxe the nightly version? Or do you have two Haxe installs?

**Comment by tobil4sk** [comment]
Yes, `/usr/local/share/haxe` is from Haxe nightly. I only have one version installed at a time.

**Comment by joshtynjala** [comment]
> /usr/local/share/haxe/std/haxe/display/Protocol.hx:32: characters 2-105 : Inline variable initialization must be a constant value

Display.hx and Protocol.hx are part of the Haxe standard library. If the Haxe nightly compiler is producing errors in its own standard library, that seems like a bigger issue, and I personally wouldn't trust any errors it reports about Lime until that issue is resolved.

**Comment by dimensionscape** [comment]
It's a nightly build....I'm going to go out on a limb here and say that this has nothing to do with lime. There are errors here in the haxe standard library. Lets avoid writing issue reports for compatibility with unofficial versions of haxe. I suggest we close this...

**Comment by player-03** [comment]
If Lime's errors were reported first, I might be worried. But since they come after a slew of standard library errors, there's a good chance they'll go away when those are fixed.

**Comment by dimensionscape** [comment]
Well, just as a rule of thumb... it's better not to jump to conclusions about compatibility with a development build of Haxe. If anything, bring an issue up over at the Haxe git in regards to this.

**Comment by player-03** [comment]
`HaxeRequestMethod` is an abstract wrapping `String`. Presumably it was allowed because static inline strings are allowed.

**Reply by player-03** [reply]
Pretty sure the point of an abstract is to be the _same_ type at runtime, but with different properties and functions available at compile time. If you use `Reflect` on an instance of an abstract, you'll only see fields from the underlying type.

**Reply by dimensionscape** [reply]
No, that's not quite the point of an abstract. In actuality, an abstract is a type definition that provides a common interface for a set of related types. The purpose of an abstract is to abstract away the specific details of each related type and provide a common interface that can be used to interact with any of the related types. This enables polymorphism. 

I mean, the documentation is quite literal in this definition.

![](https://i.gyazo.com/60807d61f1dce24a383181df3da8d806.png)

**Reply by player-03** [reply]
Ok, so the confusion is the words "same" and "different." The Haxe documentation says "different" because they mean "different from the abstract type." I said "same" because I meant "the same as the underlying type." Either wording works, but I guess you gotta clarify.

- In their example, `AbstractInt` is a different type at runtime. That is, it's not `AbstractInt`, it's just `Int`.
- In my example, `HaxeRequestMethod` wraps `String`. At runtime, the static variables are still `String`s, the same type as before they were wrapped.



> The purpose of an abstract is to abstract away the specific details of each related type and provide a common interface that can be used to interact with any of the related types. This enables polymorphism.

That's an interface or abstract class, not an abstract. Isn't it great that Haxe named two unrelated language features nearly identical names?

**Reply by player-03** [reply]
> The whole point of abstracts is that it is a blueprint designed to be a different type at runtime, which defies the definition of an inlined, compile time constant.

Another way to think about it: if the underlying type qualifies as a constant value, the abstract type does too. So you can define new constant values by abstracting old ones.

**Reply by dimensionscape** [reply]
> That's an interface or abstract class, not an abstract. Isn't it great that Haxe named two unrelated language features nearly identical names?

Both abstracts and interfaces provide a mechanism for achieving polymorphism, but they work in slightly different ways...

As I'm sure we both know, an interface is a contract that a class can choose to implement. It defines a set of methods, fields, and properties that the class must implement. 

Conversely, Abstracts in Haxe enable us sort of an abstract polymorphism. 

I wrote an example, which you can test here: https://try.haxe.org/#2Cb8C87C

```hx
@:forward
abstract Person(String) to String {
	public var name(get, set):String;
	public var surname(get, set):String;
	public var age(get, set):Int;

	public function new(name:String, surname:String, age:Int) {
		this = name + "::" + surname + "::" + Std.string(age);
	}

	public inline function get_name():String {
		return __split(this)[0];
	}

	public inline function set_name(value:String):String {
		return this = __modify(this, 0, value);
	}

	public inline function get_surname():String {
		return __split(this)[1];
	}

	public inline function set_surname(value:String):String {
		this = __modify(this, 1, value);
		return value;
	}

	public inline function get_age():Int {
		return Std.parseInt(__split(this)[2]);
	}

	public inline function set_age(value:Int):Int {
		this = __modify(this, 2, Std.string(value));
		return value;
	}

	@:to
	private inline function toObject():Dynamic {
		var s:String = this;
		var arr:Array<String> = s.split("::");

		return {name: arr[0], surname: arr[1], age: arr[2]};
	}

	private static function __split(s:String):Array<String> {
		return s.split("::");
	}

	private static function __modify(s:String, index:Int, x:String):String {
		var arr:Array<String> = __split(s);
		arr[index] = x;

		return arr.join("::");
	}
}
```

Yes, it's true that the underlying definition is a String, but we've outlined new variables as well. At runtime, it is no longer a String but a Person Type, complete with a unique constructor, and it's own fields. Not only that, it can behave as a String or an Object with definitions for its implicit transition to an Object.

Here is the output from tracing out it's unique fields, as a String, and as an Object from the try.haxe example:
```
18:08:00:914   Test.hx:4: 34, Chris, Speciale
18:08:00:914   Test.hx:7: Chris::Speciale::34
18:08:00:914   Test.hx:11: 34, Chris, Speciale
18:08:00:914   Test.hx:13: { name : Chris, surname : Speciale, age : 34 }
```

**Reply by dimensionscape** [reply]
Not only that, but an Abstract can use an Interface as its underlying Type in order to provide powerful inheritance based architecture for Anonymous Structures like this:

```hx
@:forward
@:transitive
abstract LoginMessage(Message<ILoginMessage>) from ILoginMessage
{
	public static inline var LOGIN_STATUS:MessageType<ILoginMessage> = "loginStatus";

	public inline function new(){
		this = new Message<ILoginMessage>();
		this.data = new Object();
	}
}

@:forward
@:transitive
abstract Message<T>(T) from T to T to Object 
{
	public inline function new(){
		this = new Object();
	}
}

interface ILoginMessage extends IDataMessage<ILoginData>
{
	
}

interface ILoginData 
{
	var username:String;
	var password:String;

}

interface IDataMessage<T> extends IMessage
{
	public var data:T;	
}

interface IMessage 
{
	public var type:String;
	public var connection(default, null):Connection;
}
```

**Comment by dimensionscape** [comment]
Ok, nvm, I guess it makes sense if the Abstract is fully resolvable at compile-time. That's actually not that head-spinning. It has nothing to do with the underlying type per se, only that it can be completely resolved during compilation.

**Reply by player-03** [reply]
Ah cool, glad we got this figured out.

I'm going to post all the stuff I was typing up, but now that we're on the same page you don't have to respond to it.

> if the Abstract is fully resolvable at compile-time.

Well, `Reflect` can't access the `name`, `surname`, and `age` properties, so all of the fields pretty much have to be resolved at compile time.

I say "pretty much" because abstract implementation classes _are_ a thing. (They always show up with `-Ddump=pretty`, though with DCE enabled they might be empty.) If you disabled DCE and got an instance of one of those, I bet `Reflect` could access everything. But I'm not sure how you'd get that instance.

> At runtime, it is no longer a String but a Person Type,

~~Again, that's compile time. Just like static extensions only apply at compile time, and for the same reasons. Just like typedefs change the compile-time type but not the runtime type.~~

~~Maybe each of us has a different idea of what it means to "be something at runtime"?~~

The way I see it, `Context.typeof()` (during a macro) gets the compile-time type, and `Type.getClass()` gets the runtime type. And if you try `Type.getClassName(Type.getClass(person))`, you get `String`.

> ~~A final could make more sense here(not in my example, but in general).~~

[The Haxe devs agree](https://haxe.org/manual/class-field-inline.html#trivia-inline-var). Hopefully they'll go ahead with the deprecation.

And IMO it could make sense for `Person` to be constant, in the context of a game where you know for sure none of the character's names or ages will change. For real people, agreed.

**Reply by dimensionscape** [reply]
Sorry, I tend to get twisted up on these things sometimes.

My curiosity is really bugging me so I filed an issue: https://github.com/HaxeFoundation/haxe/issues/11005#issue-1612434297

> And IMO it could make sense for Person to be constant, in the context of a game where you know for sure none of the character's names or ages will change. For real people, agreed.

Yeah it could be, if you knew what the value was before runtime. I guess my point there was that the underlying String value didn't necessarily make it eligible to be a constant. The age, name, and surname fields could be read only, but in this case they're read/write, disqualifying it from being one. 

Anyway, nice discussion. 🤣

**Reply by player-03** [reply]
> The age, name, and surname fields could be read only, but in this case they're read/write, disqualifying it from being one.

It actually compiles as long as you don't try to set those values. And if you do try it prints an error message specific to this exact situation. Which makes me think "abstracts as constants" are intended and have been around for some time.

Edit: Actually, you also need to inline the constructor. Then it'll compile.

**Reply by dimensionscape** [reply]
> > The age, name, and surname fields could be read only, but in this case they're read/write, disqualifying it from being one.
> 
> It actually compiles as long as you don't try to set those values. And if you do try it prints an error message specific to this exact situation. Which makes me think "abstracts as constants" are intended and have been around for some time.

Yeah, obviously they have been, since core types like UInt are abstract. It still seems a bit weird to me personally, although I wonder if they changed the way this behaves due to the errors noted before.

On a side note I don't believe final and inline quite do the same thing at the moment irrc in tests looking at certain use cases in transpiled code. Often they do the same thing, but sometimes they don't, unless there have been changes in recent versions I'm unaware of. 

Here are a few differences I noticed:

1)
When accessing the value from different modules: If you declare a static final value in one module and access it from another module, the value will be accessed via a function call at runtime. On the other hand, if you use static inline, the value will be inlined and accessed directly, resulting in faster execution.

2)
When the value is used in a loop: If you use a static final value in a loop, it will be accessed via a function call on each iteration, which can lead to a performance overhead. However, if you use static inline, the value will be inlined and accessed directly, resulting in faster execution.

3)
When the value is used in an expression: If you use a static final value in an expression, it may not be inlined by the Haxe compiler, depending on the context. However, if you use static inline, the value will be inlined, resulting in faster execution.

**Reply by player-03** [reply]
> On a side note I don't believe final and inline quite do the same thing

They don't, and I didn't mean to imply otherwise. In my comment above, I was talking about `static inline final`, not `static final`.

I distinctly recall mentioning the fact that `inline final` is the same as `inline var`. I know I used those specific phrases, but neither phrase shows up anymore. I guess I cut that part during editing? Whoops.

**Reply by dimensionscape** [reply]
> > On a side note I don't believe final and inline quite do the same thing
> 
> They don't, and I didn't mean to imply otherwise. In my comment above, I was talking about `static inline final`, not `static final`.
> 
> I distinctly recall mentioning the fact that `inline final` is the same as `inline var`. I know I used those specific phrases, but neither phrase shows up anymore. I guess I cut that part during editing? Whoops.

Oh, I was actually just reflecting on my own thoughts. It was a little confusing at one time a while ago.

**Comment by Apprentice-Alchemist** [comment]
Removing `-D no-inline` from `tools/tools.hxml` fixes the issue.

**Reply by tobil4sk** [reply]
Looks like this was done to fix some old Haxe 3.2.1 issue... https://github.com/openfl/lime/commit/14f064d33aabf9dd1622fa202997055e4f54f830

**Reply by Apprentice-Alchemist** [reply]
> Firstly, why would we want to prevent inlines in Lime in the first place? Seems kind of absurd right? And under what context? Are we trying to minimize code vs performance? Strange.

Note that this only applies to the tools, not lime in general.

> why is no-inline suddenly breaking abstract inlines

If I had to guess I'd "blame" https://github.com/HaxeFoundation/haxe/pull/10858, since it adds `import haxe.display.Display` to `haxe.macro.Compiler`.

**Reply by dimensionscape** [reply]
Oh I see, this is only used in building lime itself.

**Reply by player-03** [reply]
> Second, why is no-inline suddenly breaking abstract inlines?

Hmm. I mentioned earlier that inlining the constructor is required to use the value as a static inline. Otherwise it doesn't think it's a constant value.

And normally I assume no-inline is fine because it removes the `inline` keyword from both the constructor and the static variable. But maybe ~~in the nightly version~~ it leaves the static inline? (Just speculating here.)

---

Pretty sure [this other no-inline bug](https://github.com/HaxeFoundation/haxe/issues/10912) is unrelated, but it's interesting how the same feature breaks so much. I suspect it isn't very well-tested.

**Reply by tobil4sk** [reply]
> If I had to guess I'd "blame" https://github.com/HaxeFoundation/haxe/pull/10858, since it adds import haxe.display.Display to haxe.macro.Compiler.

Yep, I checked this to confirm and this is is the first commit where `lime rebuild tools` fails.

**Reply by dimensionscape** [reply]
> And normally I assume no-inline is fine because it removes the inline keyword from both the constructor and the static variable. 

Yeah precisely. It's pretty counter intuitive to get an error about inlines when using a macro to remove inlines. 🤣

**Reply by dimensionscape** [reply]
> > If I had to guess I'd "blame" [HaxeFoundation/haxe#10858](https://github.com/HaxeFoundation/haxe/pull/10858), since it adds import haxe.display.Display to haxe.macro.Compiler.
> 
> Yep, I checked this to confirm and this is is the first commit where `lime rebuild tools` fails.

Awesome, thanks for confirming that this is indeed a haxe bug! Proud to have so many intelligent and capable developers involved in the Lime ecosystem ❤️.

On that note, what do you all think the impact on the lime toolchain is if we just remove it anyway? Are we inlining much to begin with?

**Reply by Apprentice-Alchemist** [reply]
Testing reveals that `static inline var FOO = new Foo()` with Foo an abstract with inlined constructor, and `-D no-inline` broke between 4.1.5 and 4.2.0.

While this is technically a compiler bug it's probably best to remove the flag anyway, since it could lead to other issues in the future (given that it's clearly not tested very well).

**Reply by tobil4sk** [reply]
> On that note, what do you all think the impact on the lime toolchain is if we just remove it anyway? Are we inlining much to begin with?

Since `no-inline` was added to fix what was most likely just a compilation issue in Haxe 3.2.1 in 14f064d33aabf9dd1622fa202997055e4f54f830, and this will only affect lime tools anyway, I would say the impact would be minimal. There is not much information in that commit message, so if we want to be 100% safe we could try to compile that commit with 3.2.1. If compilation fails without no-inline, then it's unlikely it was added for any reason other than to fix a compilation bug.

**Reply by dimensionscape** [reply]
Well, I think the consensus is that we aren't supporting 3.2.1 anymore. At some point its fair to stop supporting Haxe 3 all together...although that is entirely a topic for a totally different discussion. Anyways, I think @Apprentice-Alchemist is right so we should just go ahead and remove it.
